core/src/main/java/io/fabric8/maven/core/service/PortForwardService.java

```diff
@@ -0,0 +1,242 @@
+     */
+    public Closeable forwardPortAsync(final Logger externalProcessLogger, final LabelSelector podSelector, final int remotePort, final int localPort) throws Fabric8ServiceException {
+
+        final Lock monitor = new ReentrantLock(true);
+        final Condition podChanged = monitor.newCondition();
+        final LinkedList<Pod> forwardedPods = new LinkedList<>();
+
+        final Thread forwarderThread = new Thread() {
+            @Override
+            public void run() {
+
+                Pod currentPod = null;
+                Closeable currentPortForward = null;
+
+                try {
+                    monitor.lock();
+
+                    while (true) {
+                        if (forwardedPods.isEmpty() || podEquals(currentPod, forwardedPods.getLast())) {
+                            podChanged.await();
+                        } else {
+                            Pod nextPod = forwardedPods.getLast(); // may be null
+                            monitor.unlock();
+
+                            if (currentPortForward != null) {
+                                log.info("Closing port-forward from pod %s", KubernetesHelper.getName(currentPod));
+                                currentPortForward.close();
+                                currentPortForward = null;
+                            }
+
+                            if (nextPod != null) {
+                                log.info("Starting port-forward to pod %s", KubernetesHelper.getName(nextPod));
+                                currentPortForward = forwardPortAsync(externalProcessLogger, KubernetesHelper.getName(nextPod), remotePort, localPort);
+                            } else {
+                                log.info("Waiting for a pod to become ready before starting port-forward");
+                            }
+                            currentPod = nextPod;
+
+                            monitor.lock();
+                        }
+
+                    }
+
+                } catch (InterruptedException e) {
+                    log.debug("Port-forwarding thread interrupted", e);
+                    Thread.currentThread().interrupt();
+                } catch (Exception e) {
+                    log.warn("Error while port-forwarding to pod", e);
+                } finally {
+                    monitor.unlock();
+
+                    if (currentPortForward != null) {
+                        try {
+                            currentPortForward.close();
+                        } catch (Exception e) {}
+                    }
+                }
+            }
+        };
+
+        // Adding the current pod if present to the list
+        Pod newPod = getNewestPod(podSelector);
+        if (newPod != null) {
+            forwardedPods.add(newPod);
+        }
+
+
+        final Watch watch = kubernetes.pods().watch(new Watcher<Pod>() {
+
+            @Override
+            public void eventReceived(Action action, Pod pod) {
```

Comment:
I've checked.. it's for pods in the current namespace. Btw I've changed the list into a reference, since I'm always using the last value only.